### PR TITLE
refactor(nmap): migrate scanner wrapper to Ullaakut/nmap v4

### PR DESCRIPTION
**Pull request by vyskocilm**
Migrates the codebase from `github.com/Ullaakut/nmap/v3` v3.1.0 to `github.com/Ullaakut/nmap/v4` v4.0.1. After merging post-#183 main (which already pins `nmap/v4 v4.0.1` and `kaptinlin/jsonschema v0.9.3`), the net go.mod change of this PR is dropping the `Ullaakut/nmap/v3 v3.1.0` require — v4 becomes the single, actually-imported nmap dependency.

Closes #185. Supersedes Renovate #162.

## Migration

Only one production file changed: `internal/nmap/nmap.go`, three mechanical edits forced by v4's API:
- `NewScanner(ctx, options...)` → `NewScanner(options...)` (ctx moved out of the constructor)
- `scanner.Run()` returning `(*Run, *[]string, error)` → `scanner.Run(ctx)` returning `(*Run, error)`
- warnings read via `run.Warnings()` — loop kept **after** the empty-hosts early return, preserving the warnings-dropped-on-empty-runs quirk

Everything else (option flag emission, XML struct tags, `ErrScanTimeout`/`ErrScanInterrupt` ctx semantics) verified identical between v3.1.0 and v4.0.1. `nmap_test.go` and `internal/bom/golden_test.go` are pure alias swaps; golden testdata byte-identical.

## Test coverage: internal/nmap 52.9% → 100% (-short), 85.6% → 100% (full)

New deterministic suite (`scan_test.go` + XML fixtures + fake-nmap shell script, Windows-skipped, no network/binary needed in `-short`):
- exact-argv behavior locks (written and green against v3 **before** the flip, unchanged after — cross-version parity proof)
- TLS cipher/cert/SSH-hostkey parsing from fixture spliced out of a real scan capture (raw XML entities intact — exercises the `html.UnescapeString` path)
- empty-run, stderr-warnings (slog-captured, asserts warning logged for non-empty runs only), exit-error wrapping, missing-binary, builder value semantics, parse-branch units
- deep-nesting BOM validation regression (see jsonschema notes)

## Review process (subagentic: architect → developer → reviewer → tester + dedicated jsonschema audit, all max-effort)

- **Reviewer: APPROVE.** All constraints verified empirically: behavior-lock tests run at pre-flip commit (v3) and HEAD (v4) both green; `html.UnescapeString` mutation-tested (removal fails 2 tests); nil-deref safety against v4's `(nil, nil)` returns confirmed; otel v1.43.0 trio + `cenkalti/backoff/v5` verified genuinely MVS-forced by nmap v4.0.1 (`go mod graph`); `go mod tidy -diff` clean.
- **Tester: PASS.** Coverage claims reproduced; mutation kill rate **7/7** (incl. option-order, `-6`, default-ports, unknown-script, error-wrap, warnings-quirk mutants); ETXTBSY retry proven to de-flake a real fork/exec race (golang.org/issue/22315) and unable to mask genuine failures; suite passes with PATH containing only `cat`.
- **jsonschema 0.7.5 → 0.9.3 audit: APPROVE** (the bump itself shipped in #183; the audit here is the rationale for this PR's `TestValidator_DeepNesting`). 12 dual-version probes verdict-identical except one: **0.9.3 fixes a real false-pass** — invalid content at component nesting depth ≥ 5 was silently accepted by 0.7.5's depth-heuristic cycle breaker; 0.9.3 rejects it (locked by the new regression test). Offline-strictness, `SetAssertFormat(true)`, idn-email override, and `ToList` error shape all verified intact.

## Follow-ups filed during this work

- #186 — `lens.go:263` discards `WithNmapBinary`'s value-receiver return: configured `nmap.binary` silently ignored (pre-existing; found by the architect)
- #184 — migrate `docker/docker` → `moby/moby/client` (unrelated to this PR)

🤖 Generated with [Claude Code](https://claude.com/claude-code)

## v3→v4 delta addendum (from review)

v4 flipped error precedence: `finalizeRun` returns `mapRunError(runErr)` before consulting the stderr sentinels, so `ErrRequiresRoot`/`ErrMallocFailed`/`ErrResolveName` now surface as a bare `exit status N`. No practical impact for cbom-lens (no sentinel consumers; connect-mode `-sV` needs no root), but noted as a known v4 behavior change rather than "identical".

### CODE REVIEW

File: internal/bom/validator_test.go [hunk 1]
```diff
@@ -300,6 +301,73 @@ func TestValidator_JSFSignature(t *testing.T) {
...
+		b, err := bom.NewBuilder(model.CBOM{Version: "1.6"})
+		require.NoError(t, err)
+		bom := b.BOM(t.Context())
+		bom.Components = &[]cdx.Component{deeplyNested(8, leaf)}
+
+		var buf bytes.Buffer
+		enc := cdx.NewBOMEncoder(&buf, cdx.BOMFileFormatJSON)
+		require.NoError(t, enc.Encode(&bom))
+
+		return validator.Validate(&bom), validator.ValidateBytes(buf.Bytes())
```
[the remark this answers is not in the document]
Comment (vyskocilm) [reply internal/bom/validator_test.go]: Fixed in d8df2cc — renamed the local to `doc`.

File: internal/nmap/scan_test.go [hunk 2]
```diff
@@ -0,0 +1,306 @@
...
+)
+
+// fakeNmap writes a POSIX-shell fake nmap into t.TempDir() that records its
+// argv (one arg per line) to a file, optionally prints a line to stderr,
+// cats the given fixture to stdout, and exits with code.
+// Skips the calling test on Windows.
+func fakeNmap(t *testing.T, fixturePath string, stderrLine string, exitCode int) (binPath, argsFile string) {
+	t.Helper()
+	if runtime.GOOS == "windows" {
+		t.Skip("fake nmap is a POSIX shell script, skipping on Windows")
```
Comment (lubomirw) [inline internal/nmap/scan_test.go]: **[Medium]** | Testing

`fakeNmap` skips on Windows, so 5 of the 10 new tests never run there: both argv parity locks, TLS/SSH parsing, empty-run and exit-error. Measured coverage drops from the claimed 100% to 79.6% for the package — and 16.7% for `scan()`, the one function holding all three migrated v4 calls. CI is unaffected (the `check-windows` job runs a registry-scan integration binary, never `go test`), but the migrated code is unverifiable on the team's own dev platform.

_Suggested:_ Replace the `/bin/sh` script with the stdlib helper-process idiom — `TestMain` already exists, so add a branch that, when an env var is set, writes `os.Args[1:]` one per line, copies the fixture to stdout, optionally writes a stderr line, and exits with the requested code. Tests then use `t.Setenv` + `WithNmapBinary(os.Args[0])`. No shell and no written executable, which also removes the ETXTBSY retry and the `%q`-as-shell-quoting hazard.

<sub>flagged by guided-pr-review and superpowers code review</sub>

**Review by lubomirw — COMMENTED** [review]
Automated multi-reviewer triage — 7 findings (COMMENT).

Four reviewers ran in parallel against `e7d512b...d8df2cc`: Copilot CLI (0 findings), `guided-pr-review`, superpowers code review, and a comment/log hygiene scan (0 findings). Findings were deduped and independently verified against the nmap v3.1.0 and v4.0.1 module sources.

Three further items are left as judgment calls rather than posted inline: the warnings-dropped-on-empty-runs quirk lock at `internal/nmap/nmap.go:124` (worth an issue reference in the test comment so the lock reads provisional rather than desired), the dead `rawPath` field that `internal/nmap/scan_test.go:241` now permanently covers, and `docs/testing-ci.md` not mentioning the new OS-gated skip or the platform-dependent coverage figure. Out of diff scope but confirmed: `cmd/cbom-lens/lens.go:263` discards `WithNmapBinary`'s returned copy, so a configured `ports.binary` is silently ignored — already filed as #186.

File: internal/nmap/scan_test.go [hunk 3]
```diff
@@ -0,0 +1,306 @@
...
+// recordedArgs reads back the argv recorded by the fakeNmap script,
+// one argument per line.
+func recordedArgs(t *testing.T, argsFile string) []string {
+	t.Helper()
+	data, err := os.ReadFile(argsFile)
+	require.NoError(t, err)
+	return strings.Split(strings.TrimSuffix(string(data), "\n"), "\n")
+}
+
+func TestScan_EmitsExpectedArgsAndParsesTLS(t *testing.T) {
```
Comment (lubomirw) [inline internal/nmap/scan_test.go:113]: **[Medium]** | Testing

Moving the context from `NewScanner(ctx, ...)` to `Run(ctx)` is the one v4 change with real runtime semantics (cancellation, deadline → `ErrScanTimeout`/`ErrScanInterrupt`), yet every new test passes an uncancelled `t.Context()`. The suite — and the 100% coverage claim — would stay green if the migration had hardcoded `Run(context.Background())` and silently stopped honouring cancellation for 1-65535 port scans.

_Suggested:_ Add a test with a fake that sleeps before `cat`, a ctx with a ~100ms deadline, and assert both that `Scan` returns promptly and that the error chain carries `errors.Is(err, nmapv4.ErrScanTimeout)`. A cancel-mid-scan variant covers `ErrScanInterrupt`.

<sub>flagged by superpowers code review</sub>

File: internal/nmap/nmap.go [hunk 4]
```diff
@@ -96,14 +96,14 @@ func (s Scanner) Scan(ctx context.Context, addr netip.Addr) (model.Nmap, error)
...
-	scanner, err := nmap.NewScanner(ctx, options...)
+	scanner, err := nmap.NewScanner(options...)
 	if err != nil {
 		return nil, fmt.Errorf("creating nmap scanner: %w", err)
 	}
 
 	now := time.Now()
 	slog.InfoContext(ctx, "scan started")
-	scan, warningsp, err := scanner.Run()
+	scan, err := scanner.Run(ctx)
```
Comment (lubomirw) [inline internal/nmap/nmap.go:106]: **[Low]** | Error handling

v4 flipped error precedence: v3's `processNmapResult` called `checkStdErr` first (`v3/nmap.go:261`), while v4's `finalizeRun` returns `mapRunError(runErr)` before ever consulting the stderr sentinel (`v4/helpers.go:40-48`). So `ErrRequiresRoot` / `ErrMallocFailed` / `ErrResolveName` now surface as a bare `exit status N`. Practical impact here is small — cbom-lens reads no nmap sentinels and connect-mode `-sV` does not require root — but it contradicts the PR body's claim that "everything else … verified identical between v3.1.0 and v4.0.1".

_Suggested:_ Record the delta in the PR body. Optionally add a test asserting the observable error for a non-zero exit, so a future upstream fix is noticed.

<sub>flagged by superpowers code review; precedence flip confirmed against both module sources</sub>

File: go.mod [hunk 5]
```diff
@@ -5,7 +5,6 @@ go 1.26.4
 require (
 	cuelang.org/go v0.16.1
 	github.com/CycloneDX/cyclonedx-go v0.10.0
-	github.com/Ullaakut/nmap/v3 v3.1.0
 	github.com/Ullaakut/nmap/v4 v4.0.1
```
Comment (lubomirw) [inline go.mod:8]: **[Low]** | Documentation

The merge-base already pins both `kaptinlin/jsonschema v0.9.3` and `Ullaakut/nmap/v4 v4.0.1` — they landed via #183. The only go.mod change in this PR is dropping the `Ullaakut/nmap/v3 v3.1.0` require, so the title's "bump jsonschema to 0.9.3", the body's "collapsed to a single v4.0.1 pin", and the whole jsonschema-audit section read as changes that are not in this diff.

_Suggested:_ Reword the body to state the jsonschema bump already landed in #183, keeping the audit notes purely as the rationale for retaining `TestValidator_DeepNesting`.

<sub>flagged by guided-pr-review</sub>

File: internal/nmap/scan_test.go [hunk 6]
```diff
@@ -0,0 +1,306 @@
...
+func TestScan_BinaryExitErrorWrapped(t *testing.T) {
+	t.Parallel()
+	garbage := filepath.Join(t.TempDir(), "garbage.txt")
+	require.NoError(t, os.WriteFile(garbage, []byte("this is not nmap XML output\n"), 0o644))
+	bin, _ := fakeNmap(t, garbage, "", 1)
+
+	got, err := scanFake(t, New().WithNmapBinary(bin).WithPorts("443"),
+		netip.MustParseAddr("127.0.0.1"))
+	require.Error(t, err)
+	require.Contains(t, err.Error(), "nmap scan services:")
```
Comment (lubomirw) [inline internal/nmap/scan_test.go]: **[Low]** | Testing

`require.Contains(t, err.Error(), "nmap scan services:")` matches the wrapper `nmap.go:87` adds to every failure — including the constructor error `TestScan_NmapMissingFromPath` already covers — so this test discriminates nothing despite being the only one meant to pin error semantics. The fake also exits 1 *and* emits garbage stdout, conflating two failure modes, and the exit-0-plus-malformed-XML path is untested even though that is where v4 changed the error identity (v3's `ErrParseOutput` sentinel → `parsing nmap XML output: <detail>`).

_Suggested:_ Tighten to `require.ErrorContains(t, err, "exit status 1")` and add a sibling case with `exitCode` 0 plus garbage stdout asserting `"parsing nmap XML output"`.

<sub>flagged by guided-pr-review and superpowers code review</sub>

File: internal/nmap/scan_test.go [hunk 7]
```diff
@@ -0,0 +1,306 @@
...
+// (see golang.org/issue/22315). The window is tiny, so a few short retries
+// make the deterministic tests immune to it. Any other outcome, success or
+// failure, is returned as-is.
+func scanFake(t *testing.T, s Scanner, addr netip.Addr) (model.Nmap, error) {
+	t.Helper()
+	var got model.Nmap
+	var err error
+	for range 20 {
+		got, err = s.Scan(t.Context(), addr)
+		if err == nil || !strings.Contains(err.Error(), "text file busy") {
```
Comment (lubomirw) [inline internal/nmap/scan_test.go]: **[Low]** | Testing

The retry gates on `strings.Contains(err.Error(), "text file busy")` — a strerror string, not a stable API. If the wording or locale ever differs, the retry silently stops working and the fork/exec flake returns with no signal that the guard broke.

_Suggested:_ Use `errors.Is(err, syscall.ETXTBSY)`. Verified: `cmd.Start()`'s error is returned unwrapped by v4 and `mapRunError`'s default branch passes it through, so the errno survives the `os.PathError` plus both `fmt.Errorf("%w")` wraps — and `syscall.ETXTBSY` is defined on Windows too, so the file still compiles cross-platform.

<sub>flagged by guided-pr-review and superpowers code review</sub>

File: internal/bom/validator_test.go [hunk 8]
```diff
@@ -300,6 +301,73 @@ func TestValidator_JSFSignature(t *testing.T) {
...
+			Name: "deep-invalid-leaf",
+		})
+
+		// The failure chain must name the leaf's type/enum violation — the
+		// bogus value exists only on the depth-8 component, so matching it
+		// proves the validator actually descended that deep instead of
+		// bailing out fail-open like v0.7.5 did.
+		for _, err := range []error{errBOM, errBytes} {
+			require.Error(t, err)
+			require.Contains(t, err.Error(), "Property 'type' does not match the schema")
```
Comment (lubomirw) [inline internal/bom/validator_test.go]: **[Low]** | Testing

`"Property 'type' does not match the schema"` is rendered from kaptinlin/jsonschema's i18n catalogue (`i18n/locales/en.json`, `property_mismatch`) — the generic message the library's own docs flag as not-to-be-consumed, and which `internal/bom/validator.go:221-228` exists to flatten past. The assertion is redundant with line 356 and breaks on any patch-bump reword or non-en locale, for reasons unrelated to recursion depth.

_Suggested:_ Drop this line and strengthen the next one to pin the pointer structurally, e.g. ``require.Regexp(t, `(/components/0){7}/type: enum: Value not-a-component-type`, err.Error())``, proving the depth-8 descent from the JSON pointer instead of relying on vendor prose.

<sub>flagged by superpowers code review</sub>

File: internal/nmap/scan_test.go (hunk 6 above)
Comment (vyskocilm) [reply internal/nmap/scan_test.go]: Fixed in 044f210 — TestScan_BinaryExitErrorWrapped now also requires "exit status 1", and new TestScan_GarbageOutputIsParseError (exit 0 + garbage stdout) requires "parsing nmap XML output".

File: internal/nmap/scan_test.go (hunk 7 above)
Comment (vyskocilm) [reply internal/nmap/scan_test.go]: Resolved structurally by 4957ddb — the helper-process fake writes no executable, so the fork/exec ETXTBSY race (and the retry, string-matched or otherwise) is gone. Your errors.Is verification was correct; our earlier Windows-compilation concern was wrong (syscall.ETXTBSY is defined there).

File: internal/nmap/scan_test.go (hunk 2 above)
Comment (vyskocilm) [reply internal/nmap/scan_test.go]: Fixed in 4957ddb — replaced the shell script with the TestMain helper-process idiom (NMAP_FAKE_* env vars, os.Executable() as the binary). Windows t.Skip removed; `GOOS=windows go vet ./internal/nmap/` clean. The ETXTBSY retry is gone entirely — no executable is written anymore.

File: internal/nmap/scan_test.go (hunk 3 above)
Comment (vyskocilm) [reply internal/nmap/scan_test.go:113]: Fixed in 2650c82 — TestScan_ContextDeadlineReturnsScanTimeout (100ms deadline vs 10s fake sleep, returns ~0.1s, errors.Is ErrScanTimeout) and TestScan_ContextCancelReturnsScanInterrupt (cancel mid-scan, errors.Is ErrScanInterrupt). Sentinels verified to survive both %w wraps.

File: internal/bom/validator_test.go (hunk 8 above)
Comment (vyskocilm) [reply internal/bom/validator_test.go]: Fixed in 3b196bf with one adjustment: the proposed (/components/0){7}/type regexp matches nothing — kaptinlin/jsonschema resets the instance pointer at every $ref hop and the formatter dedupes identical detail lines, so no accumulated pointer exists. The i18n line is dropped; assertions pin structural fragments (^/components: items:, ^/0: $ref:, ^/type: enum: Value not-a-component-type). Depth-8 descent remains proven by the value: the bogus enum exists only on the leaf, and 0.7.5 fail-open produced no error at all.

File: go.mod (hunk 5 above)
Comment (vyskocilm) [reply go.mod:8]: Reworded — title is now "refactor(nmap): migrate scanner wrapper to Ullaakut/nmap v4" and the body states the jsonschema 0.9.3 + v4.0.1 pins landed via #183, keeping the audit as rationale for TestValidator_DeepNesting.

File: internal/nmap/nmap.go (hunk 4 above)
Comment (vyskocilm) [reply internal/nmap/nmap.go:106]: Recorded in the PR body ("v3→v4 delta addendum"). Observable pin added in 044f210: exit-1 asserts "exit status 1", exit-0+garbage asserts "parsing nmap XML output" — a future upstream precedence change will surface.

**Review by lubomirw — APPROVED** [review]
